packages/protocol/test/identity/identityproxyhub.ts

```diff
@@ -0,0 +1,199 @@
+const Registry: RegistryContract = artifacts.require('Registry')
+
+const concatanateHexStrings = (...strings: string[]) => {
+  return `0x${strings.reduce((concatanated: string, s: string) => {
+    return `${concatanated}${s.slice(2)}`
+  }, '')}`
+}
+
+const computeCreate2Address = <T>(
+  salt: string,
+  deployerAddress: string,
+  Contract: Truffle.Contract<T>
+) => {
+  const hash = web3.utils.soliditySha3(
+    concatanateHexStrings(
+      '0xff',
+      deployerAddress,
+      salt,
+      // @ts-ignore
+      web3.utils.soliditySha3(Contract._json.bytecode)
+    )
+  )
+  // Skip '0x' and the first 12 bytes (so take the last 20 bytes).
+  return `0x${hash.slice(2 + 2 * 12)}`
+}
+
+contract('IdentityProxyHub', (accounts: string[]) => {
+  let identityProxyHub: IdentityProxyHubInstance
+  let identityProxyTest: IdentityProxyTestInstance
+  let attestations: MockAttestationsInstance
+  let registry: RegistryInstance
+  const identifier: string = '0x00000000000000000000000000000000000000000000000000000000babecafe'
+
+  beforeEach(async () => {
+    registry = await Registry.new()
+    await registry.initialize()
+    attestations = await MockAttestations.new()
+    await registry.setAddressFor('Attestations', attestations.address)
+
+    identityProxyHub = await IdentityProxyHub.new()
+    identityProxyTest = await IdentityProxyTest.new()
+    await identityProxyHub.setRegistry(registry.address)
+  })
+
+  describe('getIdentityProxy', () => {
+    it('returns the correct CREATE2 address', async () => {
+      const expectedAddress = computeCreate2Address(
+        identifier,
+        identityProxyHub.address,
+        IdentityProxy
+      )
+      const onchainAddress = await identityProxyHub.getIdentityProxy(identifier)
+      assert.equal(onchainAddress.toLowerCase(), expectedAddress.toLowerCase())
+    })
+  })
+
+  describe('getOrDeployIdentityProxy', () => {
+    it('returns the correct CREATE2 address', async () => {
+      const expectedAddress = computeCreate2Address(
+        identifier,
+        identityProxyHub.address,
+        IdentityProxy
+      )
+      const onchainAddress = await identityProxyHub.getOrDeployIdentityProxy.call(identifier)
+      assert.equal(onchainAddress.toLowerCase(), expectedAddress.toLowerCase())
+    })
+
+    it('returns the address of an IdentityProxy', async () => {
+      const address = await identityProxyHub.getOrDeployIdentityProxy.call(identifier)
+      await identityProxyHub.getOrDeployIdentityProxy(identifier)
+      const bytecode = await web3.eth.getCode(address)
+      // @ts-ignore _json property not declared by typechain
+      assert.equal(bytecode, IdentityProxy._json.deployedBytecode)
+    })
+  })
+
+  describe('makeCall', () => {
+    let address: string
+
+    beforeEach(async () => {
+      address = await identityProxyHub.getIdentityProxy(identifier)
+      await identityProxyHub.getOrDeployIdentityProxy(identifier)
+    })
```

Comment:
Having `beforeEach` here means the case where the relayer contract is deployed as part of `makeCall` is not tested. Testing each of the following cases with the relayer predeployed and not would be good.

Reply:
I don't think it's worth the mess of redoing all tests both ways. Do you think just removing the call to `getOrDeploy` here would be enough? That way we're never relying on the proxy having been deployed beforehand. The function itself is tested by itself in some unit tests above.

Reply:
I'd be happy with that solution 👍